### PR TITLE
Filter LSP code actions based on the requested kinds

### PR DESCRIPTION
**Pull request by Techatrix**
I've observed that Zed's implementation of [Code Actions On Format](https://zed.dev/docs/configuring-zed#code-actions-on-format) uses the [CodeActionContext.only](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#codeActionContext) parameter to request specific code action kinds from the server. The issue is that it does not filter out code actions from the response, believing that the server will do it.

The [LSP specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#codeActionContext) says that the client is responsible for filtering out unwanted code actions:

```js
/**
* Requested kind of actions to return.
*
* Actions not of this kind are filtered out by the client before being
* shown. So servers can omit computing them.
*/
only?: CodeActionKind[];
```

This PR will filter out unwanted code action on the client side.

I have initially encountered this issue because the [ZLS language server](https://github.com/zigtools/zls) (until https://github.com/zigtools/zls/pull/2087) does not filter code action based on `CodeActionContext.only` so Zed runs all received code actions even if they are explicitly disabled in the `code_actions_on_format` setting.

Release Notes:

- Fix the `code_actions_on_format` setting when used with a language server like ZLS

### CODE REVIEW

File: crates/project/src/lsp_command.rs [hunk 1]
```diff
@@ -2090,19 +2090,33 @@ impl LspCommand for GetCodeActions {
...
             .filter_map(|entry| {
-                if let lsp::CodeActionOrCommand::CodeAction(lsp_action) = entry {
-                    Some(CodeAction {
-                        server_id,
-                        range: self.range.clone(),
-                        lsp_action,
-                    })
-                } else {
-                    None
+                let lsp_action = match entry {
```
Comment (osiewicz) [inline crates/project/src/lsp_command.rs:2102]: Seems like we could use let with irrefutable pattern ("let else" syntax) here?

File: crates/project/src/lsp_command.rs [hunk 2]
```diff
@@ -2090,19 +2090,33 @@ impl LspCommand for GetCodeActions {
...
+                    lsp::CodeActionOrCommand::CodeAction(lsp_action) => lsp_action,
+                    _ => return None,
+                };
+
+                if let Some(requested_kinds) = &requested_kinds_set {
+                    if let Some(kind) = &lsp_action.kind {
+                        if !requested_kinds.contains(kind) {
+                            return None;
+                        }
+                    }
```
Comment (osiewicz) [inline crates/project/src/lsp_command.rs:2112]:
```suggestion
                if let Some((requested_kinds, kind)) = &requested_kinds_set.zip(lsp_action_kind) {
                    if !requested_kinds.contains(kind) {
                        return None;
                    }
```

File: crates/project/src/lsp_command.rs [hunk 3]
```diff
@@ -2090,19 +2090,33 @@ impl LspCommand for GetCodeActions {
...
-                        server_id,
-                        range: self.range.clone(),
-                        lsp_action,
-                    })
-                } else {
-                    None
+                let lsp_action = match entry {
+                    lsp::CodeActionOrCommand::CodeAction(lsp_action) => lsp_action,
+                    _ => return None,
+                };
```
Comment (osiewicz) [inline crates/project/src/lsp_command.rs:2105]:
```suggestion
                let lsp::CodeActionOrCommand::CodeAction(lsp_action) = entry else {
                    return None;
                };
```

**Comment by SomeoneToIgnore** [comment]
Passing by and suggesting to add a test for this.
Some examples for inspiration: https://github.com/zed-industries/zed/blob/81ca890a2a65881b4526faeb4526b9c14c1228d5/crates/project/src/project_tests.rs#L4965

**Review by osiewicz — APPROVED** [review]
LGTM

**Pull request merged**